flink/src/main/java/org/apache/iceberg/flink/IcebergTableSink.java

```diff
@@ -59,12 +60,22 @@ public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
         .map(UniqueConstraint::getColumns)
         .orElseGet(ImmutableList::of);
 
-    return (DataStreamSinkProvider) dataStream -> FlinkSink.forRowData(dataStream)
-        .tableLoader(tableLoader)
-        .tableSchema(tableSchema)
-        .equalityFieldColumns(equalityColumns)
-        .overwrite(overwrite)
-        .build();
+    return (DataStreamSinkProvider) dataStream ->  {
+      // For CDC case in FlinkSQL, change log will be rebalanced(default partition strategy) distributed to Filter opr
+      // when set job default parallelism greater than 1. That will make change log data disorder and produce a wrong
+      // result for iceberg(e.g. +U comes before -U). Here try to specific the Filter opr parallelism same as it's
+      // input to keep Filter chaining it's input and avoid rebalance.
+      Transformation<?> forwardOpr = dataStream.getTransformation();
+      if (forwardOpr.getName().equals("Filter") && forwardOpr.getInputs().size() == 1) {
```

Comment:
Correct me if I'm wrong, but can't users set the name of the operators themselves?

Is there a way (possibly in the Flink SDK) that we can determine if we're in CDC mode (or more generally, if we have a retract stream) that is more specific than this? I'm worried we might catch more cases here than we mean to and unnecessarily force parallelism on jobs that don't require it.

Reply:
The `Filter` opr is appended by flink to prevenet send null data to downstream. So the `Fliter` opr will alway exists when we use FlinkSQL to construct our job, and I think the opr name can not be change by user(so far I'm not found relevant api to rename an opr by sql).

And I think we can simply set the `Filter` opr parallelism as its input and keep `Filter` chain its input opr in all case and without worrying some side-effect. Because `Filter` will not effect the input stream.

Here only change the forward opr parallelism when forward opr is `Filter`, I think that is generally enough and safe condition to keep `Filter` chain its input

Reply:
So I'm still not comfortable with relying on this `Filter` operation being appended. This assumes that Flink won't change that behavior, or that we'll catch it in between releases.

Is there another way to detect the CDC case without relying on the `Filter` operation?

Also, I think we'll potentially catch more cases than just CDC by relying on `Filter` (as that's a pretty generic operation name).

Is there no way to detect that the source is a change data capture? Or that we more generally have a `RetractStream` (unless I'm mistaken, The +U / -U should be in play because this is a `RetractStream`).

Reply:
I guess this could also be an `UpsertStream`.

Would it make more sense to check if the stream is either an `UpsertStream` or a `RetractStream`? Checking for `Filter` seems brittle to me and as though it likely casts a very wide net.

Reply:
I found that we can capture the inputstream changelog mode when A is called, and then determine whether `Filter` is needed to chain its input by checking inputstream changelog mode.

Reply:
Resolve for https://github.com/apache/iceberg/issues/2918#issuecomment-926804817